### PR TITLE
suggestion changes to chore/parallelize-list-commit-2

### PR DESCRIPTION
**Pull request by nopcoder**

### CODE REVIEW

**Comment by itaiad200** [comment]
Resolved conflicts for checks to run

File: pkg/catalog/catalog.go [hunk 1]
```diff
@@ -1043,90 +1048,103 @@ func (c *Catalog) listCommitsWithPaths(ctx context.Context, repository *graveler
...
-					return nil
+		defer close(resultCh)
+		// read and sort by heap the result to results channel
+		current := 0
+		for result := range outCh {
+			heap.Push(&jobsHeap, result)
+			for len(jobsHeap) > 0 && jobsHeap[0].order == current {
+				job := heap.Pop(&jobsHeap).(*commitLogJob)
+				if job.log != nil {
+					resultCh <- job.log
```
Comment (itaiad200) [inline pkg/catalog/catalog.go:1115]: I think you can hang here forever if the reader (`foundAllCommits`) `break` and stopped reading from the channel.

**Review by itaiad200 — CHANGES_REQUESTED** [review]

File: pkg/catalog/catalog.go [hunk 2]
```diff
@@ -1043,90 +1048,103 @@ func (c *Catalog) listCommitsWithPaths(ctx context.Context, repository *graveler
...
-				_, foundAll := findAllCommits(results, params, createdTasks)
-				if foundAll {
-					drain(done, out)
-					return nil
+		defer close(resultCh)
+		// read and sort by heap the result to results channel
+		current := 0
+		for result := range outCh {
+			heap.Push(&jobsHeap, result)
+			for len(jobsHeap) > 0 && jobsHeap[0].order == current {
```
Comment (itaiad200) [inline pkg/catalog/catalog.go:1112]: You might get stuck forever here, the `results` channel is not ordered. There's no guarantee that `current` will be at the top of the queue.

File: pkg/catalog/catalog.go [hunk 3]
```diff
@@ -1181,51 +1199,34 @@ func logCommitsResult(commits []*CommitLog, params LogParams) ([]*CommitLog, boo
...
-	for i := 0; i < len(results); i++ {
-		log, ok := results[i]
-		if !ok {
-			// we're missing a result - not done yet
-			return nil, false
-		}
-		if log != nil {
-			commits = append(commits, log)
-		}
+type commitLogJobHeap []*commitLogJob
```
Comment (itaiad200) [inline pkg/catalog/catalog.go:1207]: Document it's a min heap

File: pkg/catalog/catalog.go (hunk 1 above)
Comment (nopcoder) [reply pkg/catalog/catalog.go:1115]: added drain loop after we process the results

**Review by itaiad200 — APPROVED** [review]

**Pull request merged**